cql3/restrictions/statement_restrictions.cc

```diff
@@ -230,7 +230,8 @@ statement_restrictions::statement_restrictions(database& db,
     process_clustering_columns_restrictions(select_a_collection, for_view, allow_filtering);
 
     // Covers indexes on the first clustering column (among others).
-    if (_is_key_range && _has_queriable_ck_index) {
+    if (_is_key_range && _has_queriable_ck_index &&
+        !dynamic_pointer_cast<multi_column_restriction>(_clustering_columns_restrictions)) {
         _uses_secondary_indexing = true;
     }
```

Comment:
On Cassandra, does  it use the index with multi-column restrictions, or does it give up and filter?

Since (a, b) = (1, 2) can be rewritten as a = 1 AND b = 2, an index on either a or b should work.

Reply:
> On Cassandra, does it use the index with multi-column restrictions, or does it give up and filter?

I can't answer that without research, but why does it matter?  Are Cassandra indices sufficiently similar to ours to make that answer applicable?
 
> Since (a, b) = (1, 2) can be rewritten as a = 1 AND b = 2, an index on either a or b should work.

How?  What index-table query would give you exactly the right rows, without further filtering?

Reply:
There are some specific cases which can be done without filtering. For example, you can do `a = 1 AND k = 2`, where `a` is an indexed column and `k` is the partition key, without filtering. Or, unrelated to indexes, you can do `k = 1 AND c = 2` for the partition and clustering key. If (k, c) = (1, 2) is just another syntax for the same thing, why should it require filtering?

Reply:
> There are some specific cases which can be done without filtering.

In all such cases the base table can be queried without filtering, too (unless I'm missing something).  And that will always be faster.

> Or, unrelated to indexes, you can do `k = 1 AND c = 2` for the partition and clustering key. If (k, c) = (1, 2)

[Only clustering columns](https://sourcegraph.com/github.com/scylladb/scylla/-/blob/cql3/multi_column_relation.hh#L230:50) can participate in a multi-column restriction; this query is invalid.

Reply:
> > On Cassandra, does it use the index with multi-column restrictions, or does it give up and filter?
> 
> I can't answer that without research, but why does it matter? Are Cassandra indices sufficiently similar to ours to make that answer applicable?
> 
> > Since (a, b) = (1, 2) can be rewritten as a = 1 AND b = 2, an index on either a or b should work.
> 
> How? What index-table query would give you exactly the right rows, without further filtering?

You would need further filtering. But presumably, if you have an index on a and you then filter for b, that is better than a full scan filtering on both.

Reply:
> if you have an index on a and you then filter for b, that is better than a full scan filtering on both.

There is neither full scan nor filtering, though -- if `(a,b)=?` is allowed at all, then `(a,b)` is a clustering prefix.

Reply:
Are tuple restrictions rejected early if they are not a clustering prefix?

Reply:
> Are tuple restrictions rejected early if they are not a clustering prefix?

Yes: see [link](https://sourcegraph.com/github.com/scylladb/scylla/-/blob/cql3/multi_column_relation.hh#L230:50) above.

Reply:
Ok.